internal/k8s/vcf_controller.go

```diff
@@ -0,0 +1,181 @@
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+package k8s
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"os"
+
+	"github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/internal/lib"
+	"github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/pkg/utils"
+
+	"github.com/vmware/alb-sdk/go/clients"
+	"github.com/vmware/alb-sdk/go/session"
+
+	corev1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	typedcorev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/record"
+)
+
+func (c *AviController) AddNCPBootstrapEventHandler(k8sinfo K8sinformers, stopCh <-chan struct{}, startSyncCh chan struct{}) {
+	cs := k8sinfo.Cs
+	utils.AviLog.Debugf("Creating event broadcaster for NCP Bootstrap CRD")
+	eventBroadcaster := record.NewBroadcaster()
+	eventBroadcaster.StartLogging(utils.AviLog.Debugf)
+	eventBroadcaster.StartRecordingToSink(&typedcorev1.EventSinkImpl{Interface: cs.CoreV1().Events("")})
+	NCPBootstrapHandler := cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			utils.AviLog.Infof("NCP Bootstrap ADD Event")
+			if c.ValidBSData() && startSyncCh != nil {
+				c.UpdateAviSecret()
+				startSyncCh <- struct{}{}
+				startSyncCh = nil
+			}
+		},
+		UpdateFunc: func(old, obj interface{}) {
+			utils.AviLog.Infof("NCP Bootstrap Update Event")
+			if c.ValidBSData() && startSyncCh != nil {
+				c.UpdateAviSecret()
+				startSyncCh <- struct{}{}
+				startSyncCh = nil
+			}
+		},
+	}
+	c.dynamicInformers.NCPBootstrapInformer.Informer().AddEventHandler(NCPBootstrapHandler)
+
+	go c.dynamicInformers.NCPBootstrapInformer.Informer().Run(stopCh)
+	if !cache.WaitForCacheSync(stopCh, c.dynamicInformers.NCPBootstrapInformer.Informer().HasSynced) {
+		runtime.HandleError(fmt.Errorf("Timed out waiting for caches to sync"))
+	} else {
+		utils.AviLog.Info("Caches synced for NCP Bootstrap informer")
+	}
+}
+
+// HandleVCF checks if avi secret used by AKO is already present. If found, the it would try to connect to
+// AVI Controller. If there is any failure, we would look at Bootstrap CR used by NCP to communicate with AKO.
+// If Bootstrap CR is not found, AKO would wait for it to be created. If the authtoken from Bootstrap CR
+// can be used to connect to the AVI Controller, then avi-secret would be created with that token.
+func (c *AviController) HandleVCF(informers K8sinformers, stopCh <-chan struct{}, ctrlCh chan struct{}) {
+	cs := c.informers.ClientSet
+	aviSecret, err := cs.CoreV1().Secrets(utils.GetAKONamespace()).Get(context.TODO(), lib.AviSecret, metav1.GetOptions{})
+	if err == nil {
+		ctrlIP := os.Getenv(utils.ENV_CTRL_IPADDRESS)
+		authToken := aviSecret.Data["authToken"]
```

Comment:
Avi secret can contain username and password combination / username and token combination. Are we assuming that avi secret will alway contain authToken? Will this affect upgrade scenario from previous ako with password to this one?

Reply:
In VCF we would always use token to connect to the controller.